scope3_methodology/ad_tech_platform/model.py

```diff
@@ -0,0 +1,382 @@
+        return None
+
+    def get_corporate_allocation_decimal_fraction(self) -> float:
+        return self.corporate_allocation_percentage / 100.0
+
+    def get_servers_processing_cookie_syncs_decimal_fraction(self) -> float:
+        return self.servers_processing_cookie_syncs_percentage / 100.0
+
+    def get_servers_processing_bid_requests_decimal_fraction(self) -> float:
+        return self.servers_processing_bid_requests_percentage / 100.0
+
+    def get_bid_request_rejection_decimal_fraction(self) -> float:
+        return self.bid_request_rejection_percentage / 100.0
+
+    def get_decimal_fraction_of_bid_requests_processed_from_ad_tech_platforms(self) -> float:
+        return self.percentage_of_bid_requests_processed_from_ad_tech_platforms / 100.0
+
+    def get_data_transfer_emissions_g_co2e_per_month(self) -> float | None:
+        if self.data_transfer_emissions_mt_co2e_per_month:
+            return self.data_transfer_emissions_mt_co2e_per_month * 1000000
+        return None
+
+    def get_server_allocation_decimal_fraction(self) -> float:
+        return self.server_allocation_percentage / 100.0
+
+    def comp_external_request_ratio(self, depth: int) -> float:
+        if (
+            self.ad_tech_platform_bid_requests_processed_billion_per_month
+            and self.bid_requests_processed_billion_per_month
+        ):
+            ratio = (
+                self.ad_tech_platform_bid_requests_processed_billion_per_month
+                / self.bid_requests_processed_billion_per_month
+            )
+            log_result("external request ratio", f"{ratio * 100.0:.1f}%", depth)
+            return ratio
+        return self.get_decimal_fraction_of_bid_requests_processed_from_ad_tech_platforms()
+
+    def comp_bid_request_size_gb(self) -> float:
+        return self.bid_request_size_in_bytes / 1024 / 1024 / 1024
+
+    def comp_data_transfer_emissions_g_co2e_per_bid_request(self, depth: int) -> float:
+        # TODO why isn't this divided by the number of bid requests
+        data_transfer_emissions_g_co2e_per_month = (
+            self.get_data_transfer_emissions_g_co2e_per_month()
+        )
+        if data_transfer_emissions_g_co2e_per_month:
+            return data_transfer_emissions_g_co2e_per_month
+
+        data_transfer_emissions_g_co2e_per_bid_request = (
+            self.comp_external_request_ratio(depth - 1)
+            * self.comp_bid_request_size_gb()
+            * self.server_to_server_emissions_g_co2e_per_gb
+        )
+        log_result(
+            "data transfer emissions g co2e per bid request",
+            f"{data_transfer_emissions_g_co2e_per_bid_request:.8f}",
+            depth,
+        )
+        return data_transfer_emissions_g_co2e_per_bid_request
+
+    def comp_server_emissions_g_co2e_per_bid_request(
+        self,
+        depth: int,
+    ) -> float:
+        server_emissions_g_co2e_per_bid_request = (
+            self.get_server_allocation_decimal_fraction()
```

Comment:
just reading code - is this allocation of parent org servers? Might be worth making that more clear